Fixed #12990, Refs #27694 -- Added JSONField model field.

laymonage opened the pull request:
Following #11452.
Ticket [#12990](https://code.djangoproject.com/ticket/12990), as part of the [Google Summer of Code](https://g.co/gsoc) program.

Some points:
- Currently supports storing and retrieving any valid JSON value (boolean, integer, float, string, object, array) on all supported database backends (SQLite, PostgreSQL, MySQL, MariaDB, Oracle).
  Note: Oracle only supports JSON object and array if `IS JSON` constraint is enabled.
- Supports custom encoder and decoder (defaults to `json.JSONEncoder` and `json.JSONDecoder`, respectively). I had to cast the `SELECT` query to `text` for PostgreSQL to avoid having the value already decoded in `from_db_value` so it's possible to use a custom decoder (see [psycopg2 docs](http://initd.org/psycopg/docs/extras.html#json-adaptation)).
- Custom lookups and transforms from `contrib.postgres.fields.JSONField` are implemented as possibly supported for each backend.
- Check constraints are applied.  
  - MariaDB and SQLite have a `JSON_VALID` function which works well.
  - Oracle has [`IS JSON`](https://docs.oracle.com/en/database/oracle/oracle-database/12.2/adjsn/conditions-is-json-and-is-not-json.html#GUID-1B6CFFBE-85FE-41DD-BA14-DD1DE73EAB20) constraint, but from what I tried, it only returns true if the field's value is either a JSON object or array.
  - The constraint already comes with the JSON data types in PostgreSQL and MySQL. You can only insert valid JSON values for `jsonb` and `json` columns.
- Oracle implementation uses the `nclob` data type. Oracle recommends using `blob`, but there are some downsides (see [docs](https://docs.oracle.com/en/database/oracle/oracle-database/12.2/adjsn/overview-of-storage-and-management-of-JSON-data.html)). Besides, an existing implementation of [oracle-json-field](https://github.com/Exscientia/oracle-json-field) uses `clob`. `TextField` also uses `nclob`, so I think this makes sense.
- Needs [JSON1](https://www.sqlite.org/json1.html) extension for SQLite. Most Linux distros already have it enabled by default in their SQLite and Python packages.
- Supports introspection.

More details of this available on [my blog](https://gsoc.laymonage.com).

To test this, simply

```bash
$ pip install --no-cache --upgrade https://github.com/django/django/archive/pull/12392/head.zip
```

(or change `Django==x.y.z` in your `requirements.txt` with the link), and

```py
from django.db import models

class MyModel(models.Model):
    data = models.JSONField()
```

carltongibson commented:
Time to test this on your projects folks. 

#### PostgreSQL
Clone your DB etc. Change your Postgres JSONFields to the new `models.JSONField`. Create a migration. Migrate. Run your test suite. Did anything break? 

#### Others
Add a JSONField model. Give it a run. What doesn't work? 

Thanks all!

BallisticPain commented:
I've been working to upgrade an old codebase and have finally gotten it to Django 1.11.latest, how would I test this on my codebase? Additionally, is it a safe assumption this would be merged before Django 1.11 is EOL?

laymonage commented:
> I've been working to upgrade an old codebase and have finally gotten it to Django 1.11.latest, how would I test this on my codebase? Additionally, is it a safe assumption this would be merged before Django 1.11 is EOL?

This PR is based on the current master branch of Django, which will become 3.1. This feature will not be backported (which means it'll only be available in 3.1 and up). If you want to test this, I guess the safest way would be to make sure that your codebase can be run on 3.0, then replace 3.0 with this PR branch.

Django 1.11 is EOL in April, while this will (hopefully) be merged into 3.1 which will be released in August. The feature freeze is on May 11 though, so this should be merged before that date. See the [timeline](https://www.djangoproject.com/download/#supported-versions) and [3.1 roadmap](https://code.djangoproject.com/wiki/Version3.1Roadmap) for details.

BallisticPain commented:
> > I've been working to upgrade an old codebase and have finally gotten it to Django 1.11.latest, how would I test this on my codebase? Additionally, is it a safe assumption this would be merged before Django 1.11 is EOL?
> 
> This PR is based on the current master branch of Django, which will become 3.1. This feature will not be backported (which means it'll only be available in 3.1 and up). If you want to test this, I guess the safest way would be to make sure that your codebase can be run on 3.0, then replace 3.0 with this PR branch.
> 
> Django 1.11 is EOL in April, while this will (hopefully) be merged into 3.1 which will be released in August. The feature freeze is on May 11 though, so this should be merged before that date. See the [timeline](https://www.djangoproject.com/download/#supported-versions) and [3.1 roadmap](https://code.djangoproject.com/wiki/Version3.1Roadmap) for details.

Thank you for your prompt response. I have to move to Python 3 as my next upgrade, then I can move onward to Django 2.x -> 3.x. However, I'm implementing features and adding tests along the way so it's been slow going because I have also been learning Python and Django as I go which has been a bumpy but rewarding ride.

Thank you for the information regarding the timeline and how I could test this feature!

laymonage commented:
No problem. Good luck, and please report back once you've got there! :smile:

Taoup commented on django/db/backends/base/features.py:
```diff
@@ -295,6 +295,13 @@ class BaseDatabaseFeatures:
 
+    # Does the backend support JSONField?
+    supports_json_field = True
+    # Does the backend support primities in JSONField?
```
a typo: primitives

felixxm commented:
@laymonage Please include 7fa1a93c6c8109010a6ff3f604fda83b604e0e97 when you'll rebase this patch.

jrief commented:
In my opinion, if on the `JSONField` no `encoder` is set, it should fallback to the `django.core.serializers.json.DjangoJSONEncoder`. Otherwise fields of type `Decimal` can't be used as values inside the dictionary.
This also is the behaviour of the [jsonfield](https://pypi.org/project/jsonfield/) and [jsonfield2](https://pypi.org/project/jsonfield2/).

charettes commented:
@jrief I would argue that using `DjangoJSONEncoder` provides a false sense of security as there's no equivalent JSON types for [the ones `DjangoJSONEncoder` uses](https://github.com/django/django/blob/72b97a5b1e22f5d464045be2e33f0436fa8061d3/django/core/serializers/json.py#L83-L102) which makes the operation non-bijective.

That would result in being able to store `Decimal` objects and friends without issue but receive them back as `str` because of the codomain collisions that `DjangoJSONEncoder` causes.

adamchainz commented on django/forms/fields.py:
```diff
@@ -1211,3 +1212,65 @@ def to_python(self, value):
+    widget = Textarea
+
+    def __init__(self, encoder=None, decoder=None, **kwargs):
+        self.encoder, self.decoder = encoder, decoder
```
Nitpick : can we remove the parallel assignment here? I have a draft blog post on why it's slower (tuple construction just to immediately deconstruct), and less clear (if an exception is raised, unknown which variable caused it)

uri-rodberg commented:
> Time to test this on your projects folks.
> 
> #### PostgreSQL:
> Clone your DB etc. Change your Postgres JSONFields to the new `models.JSONField`. Create a migration. Migrate. Run your test suite. Did anything break?
> 
> #### Other Databases:
> Add a JSONField model. Give it a run. What doesn't work?
> 
> Thanks all!

Hi,

We are using Django 2.1 and we import `JSONField` from `django.contrib.postgres.fields`. I'm not sure if and how we can test the new `JSONField`? Can we test it with Django 2.1 and how? If it's just a small change I can create a branch with the new field and test it.

Siecje commented:
@uri-rodberg You need to update to the version of Django in this branch (built on the master Django branch). 

If you are on Django 2.1 you will need to update to Django 2.2 and then Django 3.0 and then this version of Django.

To test it look at this comment on this PR. https://github.com/django/django/pull/12392#issuecomment-580283664

uri-rodberg commented:
> @uri-rodberg You need to update to the version of Django in this branch (built on the master Django branch).
> 
> If you are on Django 2.1 you will need to update to Django 2.2 and then Django 3.0 and then this version of Django.
> 
> To test it look at this comment on this PR. [#12392 (comment)](https://github.com/django/django/pull/12392#issuecomment-580283664)

I can only test with Django 2.1 or 2.2 cherry-picked with the relevant commits. And I don't have the time to create such a branch. If you create such a branch that I can install it from pip, and let me know how to change the code, I can test it.

felixxm commented:
I rebased from master (with `test_key_escape()`) and made `JSONField` checks support databases aware (follow up to 71756bdfed5029bd14dce8cb3c5629efc4be55ac).

charettes commented:
@felixxm any thoughts about moving `Model._check_json_fields` to `JSONField.check` instead? 

Also it feels like [the error should be namespaced under`fields` instead of `models`](https://docs.djangoproject.com/en/3.0/ref/checks/#model-fields).

laymonage commented:
> @felixxm any thoughts about moving `Model._check_json_fields` to `JSONField.check` instead?

I think it was [originally in `JSONField.check`](https://github.com/django/django/pull/11452/commits/db19f7f640bd02b81beb95cc4fa899b412cdeb47). I'm not sure why it was moved, though.

> Also it feels like [the error should be namespaced under`fields` instead of `models`](https://docs.djangoproject.com/en/3.0/ref/checks/#model-fields).

That does make sense.

laymonage commented:
@felixxm How should we instruct users to migrate from `contrib.postgres`? Should they edit their migrations manually so it doesn't `import django.contrib.postgres.fields.jsonb.JSONField` anymore? We're not overriding `deconstruct`, so `makemigrations` has no effect. If we do, `makemigrations` will produce a migration to `models.JSONField` but that doesn't have any effect when applied either, because there's no difference in the database between both fields (which is why you asked me to remove the override).

Anyway, existing users will still need to change any references to `django.contrib.postgres.fields.jsonb.JSONField` into `models.JSONField` in their migrations (or regenerate them) when they upgrade to 4.0 to avoid import errors. 

Those who create their model with `django.contrib.postgres.fields.jsonb.JSONField` in 3.1 will have their migrations point to `models.JSONField`, so this shouldn't be a problem.

rsalmaso commented:
Run a quick test with our product (postgresql) and need to upgrade only django-filter :sweat_smile:

felixxm commented:
> @felixxm How should we instruct users to migrate from `contrib.postgres`? Should they edit their migrations manually so it doesn't `import django.contrib.postgres.fields.jsonb.JSONField` anymore? 

It's already explained in the release notes that `django.contrib.postgres.fields.jsonb.JSONField` is deprecated in favor of `django.db.models.JSONField` and will be removed in Django 4.0.

> We're not overriding `deconstruct`, so `makemigrations` has no effect. If we do, `makemigrations` will produce migrations to `models.JSONField` but they don't have any effect when applied either, because there's no difference in the database between both fields (which is why you asked me to remove the override).

Yes, and that's an expected behavior, we shouldn't generate unnecessary migrations.

felixxm commented:
> @felixxm any thoughts about moving `Model._check_json_fields` to `JSONField.check` instead?
> 
> Also it feels like [the error should be namespaced under`fields` instead of `models`](https://docs.djangoproject.com/en/3.0/ref/checks/#model-fields).

Agreed :+1:

laymonage commented:
> > @felixxm any thoughts about moving `Model._check_json_fields` to `JSONField.check` instead?
> > Also it feels like [the error should be namespaced under`fields` instead of `models`](https://docs.djangoproject.com/en/3.0/ref/checks/#model-fields).
> 
> Agreed

What error code should be used? There's a jump between `E170` and `E900`. Should I use `E903`, 'E180`, or something else?

Also, what about moving it to `JSONField.check`?

felixxm commented:
>What error code should be used? There's a jump between E170 and E900. Should I use E903, E180, or something else?

`fields.E180`

> Also, what about moving it to JSONField.check?

I agreed with the both sentences :smile: , just remember to keep it databases aware.

laymonage commented:
> > What error code should be used? There's a jump between E170 and E900. Should I use E903, E180, or something else?
> 
> `fields.E180`
> 
> > Also, what about moving it to JSONField.check?
> 
> I agreed with the both sentences smile , just remember to keep it databases aware.

If we move it back into `JSONField.check`, we won't be able to make use of `router.allow_migrate_model`, no? Therefore it will require all database connections to support `JSONField`, regardless of what `allow_migrate()` returns. I think this was why you moved it into `Model`.

charettes commented:
@laymonage you can base it off `databases` now just like `Model._check_constraints` does by relying on `self.model`.

https://github.com/django/django/blob/98f23a8af0be7e87535426c5c83058e2682bfdf8/django/db/models/base.py#L1840

felixxm commented:
I rebased from master, squashed recent commits, and pushed small edits.

I'll prepare the final edits in the next few days :crossed_fingers:.

felixxm commented:
buildbot, test on oracle.

felixxm commented:
I pushed small edits, added `KeyTransform`, `KeyTextTransform` to `docs/internals/deprecation.txt`, and moved notes about (re)moving undocumented classes in release notes. I'm still working on remaining changes.

felixxm commented:
I simplified `DataContains.as_oracle()` with `JSON_QUERY(... WITH WRAPPER)` (see [docs](https://docs.oracle.com/en/database/oracle/oracle-database/18/adjsn/clauses-used-in-functions-and-conditions-for-json.html#GUID-9079CD9C-3783-4857-9F6B-84D746C91317).

laymonage commented:
> I simplified `DataContains.as_oracle()` with `JSON_QUERY(... WITH WRAPPER)` (see [docs](https://docs.oracle.com/en/database/oracle/oracle-database/18/adjsn/clauses-used-in-functions-and-conditions-for-json.html#GUID-9079CD9C-3783-4857-9F6B-84D746C91317)).

Thanks, I guess I missed that.

felixxm commented:
I think we should restore `JSONField.empty_strings_allowed = False` (like in the original implementation) because it's unnecessary backward incompatible change (admin and migrations use this attribute). Moreover it will make `JSONField`'s behavior consistent for all backends (including Oracle). Done in bf7e79bb7ecf4a4c99870045ff41fa5e30921654.

felixxm commented:
I restored check for not callable `encoder` and added the same for not callable `decoder` (see fd31f3a137120e401b689431677da7c08b2e59ff). I also simplified some tests.

I will continue in the next week.

felixxm commented:
I rebased from master, added `DatabaseFeature.has_native_json_field` and `DatabaseOperations.json_cast_text_sql()`.

laymonage commented on django/db/backends/mysql/introspection.py:
```diff
@@ -24,6 +24,7 @@ class DatabaseIntrospection(BaseDatabaseIntrospection):
         FIELD_TYPE.DOUBLE: 'FloatField',
         FIELD_TYPE.FLOAT: 'FloatField',
         FIELD_TYPE.INT24: 'IntegerField',
+        FIELD_TYPE.JSON: 'JSONField',
```
Good call, thanks 🙂

felixxm commented:
I pushed some simplification to the `HasKeyLookup` to make it more readable (it's not the final version). I also added `test_has_key_list()` that behavior is not consistent:
- works SQLite,
- fails on PostgreSQL,
- returns wrong results on MySQL and Oracle.

felixxm commented:
I rebased from master. Unfortunately, I also confirm regression in `test_has_key_list()`.

felixxm commented on tests/model_fields/test_jsonfield.py:
```diff
@@ -0,0 +1,710 @@
+        tests = [
+            Q(value__0__has_key='a'),
+            Q(value__has_key=KeyTransform('a', KeyTransform(0, 'value'))),
+            Q(value__has_key=KeyTransform('a', KeyTransform('0', 'value'))),
```
Cases from lines 361 and 363 work with the previous implementation.

laymonage replied:
Are you sure about that? I just tried it from the latest `master` and added the following test case (basically the same test but with the model adjusted with `JSONModel`):

```py
    def test_has_key_list(self):
        obj = JSONModel.objects.create(field=[{'a': 1}, {'b': 'x'}])
        tests = [
            Q(field__0__has_key='a'),
            Q(field__has_key=KeyTransform('a', KeyTransform(0, 'field'))),
            Q(field__has_key=KeyTransform('a', KeyTransform('0', 'field'))),
        ]
        for condition in tests:
            with self.subTest(condition=condition):
                self.assertSequenceEqual(
                    JSONModel.objects.filter(condition),
                    [obj],
                )
```

The `Q(field__has_key=KeyTransform('a', KeyTransform(0, 'field')))` test failed with:

```
django.db.utils.DataError: invalid input syntax for integer: "a"
LINE 1: ...ield" ? ("postgres_tests_jsonmodel"."field" #> ARRAY[0,'a'])
                                                                  ^
```

While `Q(value__has_key=KeyTransform('a', KeyTransform('0', 'value')))` failed with:

```
django.db.utils.ProgrammingError: operator does not exist: jsonb ? jsonb
LINE 1: ...onmodel" WHERE "postgres_tests_jsonmodel"."field" ? ("postgr...
                                                             ^
HINT:  No operator matches the given name and argument type(s). You might need to add explicit type casts.
```

laymonage replied:
Also: case from line 361 still works with the new implementation. It's 362 and 363 that don't, which seem to be the case with the previous implementation as well (albeit the new implementation throws up the same error for 362 as with 363, i.e. `jsonb ? jsonb`).

Unless I'm missing something...

felixxm replied:
True, I mixed up sth when switching branches. Nevertheless they work on all databases except PostgreSQL, so we need to fix this.

felixxm commented:
>     ....
>     * returns wrong results on MySQL and Oracle.

It seems that Oracle and MySQL treat all `JSON` as arrays, that's why when we check `value__0__a='b'`, we have multiple matches:
- "unexpected" `{"a": "b"}` and
- expected `[{"a": "b"}]`.

I think we can leave it as it. Oracle and MySQL users should we aware of this caveat. I will amend `test_has_key_list()`.

uri-rodberg commented:
I can't run my app with this version of Django. I receive an error when running `make_migrations`. It's something related to the package `django-translated-fields`. It might not support this version of Django.

laymonage commented on docs/ref/databases.txt:
```diff
@@ -788,21 +788,21 @@ iterator. Your code must handle this.
-If the query doesn't throw any errors, then the JSON1 extension should already
-be enabled. Otherwise, follow the instructions below according to your
-operating system to set it up correctly.
+If the query doesn't throw any errors, then the JSON1 extension is already be
```
`is already be` -> `is already` 🙂

laymonage commented on docs/topics/db/queries.txt:
```diff
@@ -995,16 +974,9 @@ example::
 
-    Due to the limitation on Oracle and SQLite database backends, this lookup
-    is not supported.
+    This lookup is not supported on Oracle and SQLite.
```
I forgot to edit this, it's actually supported with the help of custom functions.

felixxm replied:
> I forgot to edit this, it's actually supported with the help of custom functions.

TBH, we don't have time to play with support on Oracle before the feature freeze.

laymonage replied:
Oh, I forgot to specify. I mean it's supported on SQLite 😅

felixxm commented on docs/topics/db/queries.txt:
```diff
@@ -995,16 +974,9 @@ example::
 
-    Due to the limitation on Oracle and SQLite database backends, this lookup
-    is not supported.
+    This lookup is not supported on Oracle and SQLite.
```
```
.. admonition:: Oracle

    This lookup is not supported on Oracle. 
```

It is supported on SQLite after my changes.

adamchainz commented:
> Is it not possible to backport to 2.2? It'd be a shame that this wouldn't be available in a currently supported LTS :)

No, features don't get backported. But maybe someone in the community will copy it into a third party package.

laymonage commented on docs/releases/3.1.txt:
```diff
@@ -67,16 +67,17 @@ It should have no noticeable effect on any existing Django projects.
-reference to the new ones and are deprecated as of this release.
+If your project uses ``django.contrib.postgres.fields.JSONField``, plus the
+related form field and transforms, you should adjust to use the new fields,
+and generate and apply a database migration. For now, the old fields and
```
By `generate`, do you mean running `makemigrations`? That would create unnecessary (and problematic) migrations, wouldn't it? Shouldn't the instructions be to "modify existing migrations so they point to `models.JSONField`"?

laymonage commented on docs/releases/3.1.txt:
```diff
@@ -67,16 +67,17 @@ It should have no noticeable effect on any existing Django projects.
-reference to the new ones and are deprecated as of this release.
+If your project uses ``django.contrib.postgres.fields.JSONField``, plus the
+related form field and transforms, you should adjust to use the new fields,
+and generate and apply a database migration. For now, the old fields and
```
By `generate`, do you mean running `makemigrations`? That would create unnecessary (and problematic) migrations, wouldn't it? Shouldn't the instructions be to "modify existing migrations so they point to `models.JSONField`"?

felixxm replied:
> By `generate`, do you mean running `makemigrations`?

Yes. 

> That would create unnecessary (and problematic) migrations, wouldn't it?

It will create a necessary migration to match a model state. It won't be problematic. 

> Shouldn't the instructions be to "modify existing migrations so they point to `models.JSONField`"?

No.

adamchainz replied:
> Shouldn't the instructions be to "modify existing migrations so they point to models.JSONField"?

I think they should. In django 4.0, the postgres JSONField will be dropped, so any historical migrations will fail with `ImportError` at that point, and will need updating to point to the new field anyway.

laymonage replied:
> It will create a necessary migration to match a model state. It won't be problematic.

Ah, yes I forgot about [this](https://github.com/django/django/pull/12392#issuecomment-583797410). Still, what kind of situation would `makemigrations` produce a migration in this case, though? If users are just moving from `contrib.postgres.fields.JSONField` to `models.JSONField`, `makemigrations` won't produce any migrations... (unless they change something else in the model, but that would be unrelated).

> No.

Later on, when `contrib.postgres.fields.JSONField` is removed, an `ImportError` would happen if users still have `contrib.postgres.fields.JSONField` imports in their old migrations. What should they do in that case?

felixxm replied:
> Still, what kind of situation would `makemigrations` produce a migration in this case, though? If users are just moving from `contrib.postgres.fields.JSONField` to `models.JSONField`, `makemigrations` won't produce any migrations...

It will create a migration. Have you checked this? 

> Later on, when `contrib.postgres.fields.JSONField` is removed, an `ImportError` would happen if users still have `contrib.postgres.fields.JSONField` imports in their old migrations. What should they do in that case?
> 
> Edit: ninja'd by @adamchainz

We can support it in historical migrations (like we did in the past with other fields).

felixxm replied:
> We can support it in historical migrations (like we did in the past with other fields).

I will amend it later.

laymonage replied:
> It will create a migration. Have you checked this?

Ah, correct. It didn't the last time I checked, but maybe I was doing something wrong. I just tried again and it did create a migration. Cheers :v:

felixxm replied:
Amended.

laymonage commented:
@felixxm I've pushed an initial fix for `test_has_key_list`. I also added `test_has_key_deep` because the problem wasn't about the value being a `list`, but rather the `rhs` is a chained `KeyTransform`.

I also added deep and varying levels tests for `HasKeys` and `HasAnyKeys`. However, since we're using the `?&` and `?|` shorthands on PostgreSQL, it is impossible to have varying levels of keys. On other backends, it's possible because we process each key separately and use `AND`/`OR` chains. How do you think we should approach this problem?

Right now, using `KeyTransforms` in `HasKeys` and `HasAnyKeys` errors out because psycopg2 receives `KeyTransform` objects directly. In the previous implementation, everything gets converted to `str` in `get_prep_lookup`, so it doesn't raise an error (but it doesn't work anyway, because casting `KeyTransform` to `str` doesn't make sense).

If we just want to deal with the deep lookups, I can think of assuming all keys in the `rhs` are on the same level, take one of them and move all the transforms (except the last) to the `lhs`, take the last transform of each key, put them into a list, and set it as the `rhs`. This is quite dirty. I think it's better to just not support it (even though it works on other backends) because it's not the intended way to do the query (they can always do `value__a__b__has_keys=[key1, key2, ...]`).

If we also want to deal with the varying levels lookups... well, that would mean we have to ditch `?&`/`?|` and use `AND`/`OR` instead. However, I'm afraid this is worse in terms of performance (I don't know exactly, but it's possible `?&` and `?|` are optimized). Anyway, users can always use `has_key` with `Q` and `|` if the keys are on different levels.

As a side note, PostgreSQL 12 has a [`jsonb_path_exists` function](https://www.postgresql.org/docs/current/functions-json.html#FUNCTIONS-JSON-PROCESSING-TABLE), but that is way too far from the minimum version supported by Django.

Oh, and I also switched the inheritance of `HasKeys` and `HasAnyKeys` because that's how it is in the previous implementation (I think I mixed them up initially).

laymonage commented:
@felixxm I [left a comment](https://github.com/django/django/pull/12392#issuecomment-622829720), just in case you haven't seen it since it was buried after the force-push.

laymonage commented:
Are you sure about https://github.com/django/django/pull/12392/commits/2a6b4a749d895a95a4a8fa9318309e38bf96fbdf? @carltongibson specifically asked for that back then.

felixxm commented:
> @felixxm Are you sure about [2a6b4a7](https://github.com/django/django/commit/2a6b4a749d895a95a4a8fa9318309e38bf96fbdf)? @carltongibson specifically asked for that back then.

Yes, we moved it to the wiki page.

laymonage commented on docs/topics/db/queries.txt:
```diff
@@ -892,14 +893,21 @@ array::
+    :lookup:`contains`, :lookup:`icontains`, :lookup:`endswith`,
+    :lookup:`iendswith`, :lookup:`iexact`, :lookup:`regex`, :lookup:`iregex`,
+    :lookup:`startswith`, :lookup:`istartswith`, :lookup:`lt`, :lookup:`lte`,
+    :lookup:`gt`, and :lookup:`gte` lookups.
```
Do we need to explicitly state that we can also chain them with `JSONField` lookups (e.g. `has_key`, `has_any_keys`, except `contains`)?

felixxm replied:
:thinking: How we can chain `has_key` with these lookup (e.g. `endswith`)?

laymonage replied:
No, I mean `Key, index, and path transforms` can be chained with `JSONField`-only lookups.

felixxm replied:
I don't think we need to document this. Personally, I'm happy with current docs changes.

laymonage commented on docs/ref/models/fields.txt:
```diff
@@ -1184,18 +1184,21 @@ values are stored as null.
-    not supported by the standard JSON serializer (``datetime``, ``uuid``,
-    etc.). For example, you can use the
+    not supported by the standard JSON serializer (``datetime.datetime``,
+    :class:`~python:uuid.UUID`, etc.). For example, you can use the
```
@felixxm 
You changed

```
(``datetime.datetime``, :class:`~python:uuid.UUID`, etc.).
```

into

```
(e.g. ``datetime.datetime`` or :class:`~python:uuid.UUID`).
```

in `docs/ref/forms/fields.txt`, just thought that maybe you want to do the same here.

felixxm commented:
> I also added deep and varying levels tests for `HasKeys` and `HasAnyKeys`. However, since we're using the `?&` and `?|` shorthands on PostgreSQL, it is impossible to have varying levels of keys. On other backends, it's possible because we process each key separately and use `AND`/`OR` chains. How do you think we should approach this problem?
> 
> Right now, using `KeyTransforms` in `HasKeys` and `HasAnyKeys` errors out because psycopg2 receives `KeyTransform` objects directly. In the previous implementation, everything gets converted to `str` in `get_prep_lookup`, so it doesn't raise an error (but it doesn't work anyway, because casting `KeyTransform` to `str` doesn't make sense).
> 

`has_keys` and `has_any_keys` accept list of keys and it should stay that way IMO (see https://github.com/django/django/pull/11863#issuecomment-537853174). I reverted tests.

felixxm commented on tests/model_fields/test_jsonfield.py:
```diff
@@ -393,79 +393,12 @@ def test_has_keys(self):
 
-    def test_has_keys_deep(self):
-        tests = [
-            Q(value__baz__has_keys=['a', 'c']),
```
This is covered by `test_lookups_with_key_transform()`.

felixxm commented on tests/model_fields/test_jsonfield.py:
```diff
@@ -393,79 +393,12 @@ def test_has_keys(self):
-    def test_has_any_keys_deep(self):
-        obj = NullableJSONModel.objects.create(value={'baz': {'a': 'x', 'd': 'z'}})
-        tests = [
-            Q(value__baz__has_any_keys=['c', 'd']),
```
This is covered by `test_lookups_with_key_transform()`.

laymonage commented:
> `has_keys` and `has_any_keys` accept list of keys and it should stay that way IMO (see [#11863 (comment)](https://github.com/django/django/pull/11863#issuecomment-537853174)). I reverted tests.

@felixxm
I agree.

Do you want me to do the rebase, or should I just leave it to you?

felixxm commented:
> Do you want me to do the rebase, or should I just leave it to you?

I'm working on this patch, so you can leave it to me.

laymonage commented on tests/model_fields/test_jsonfield.py:
```diff
@@ -399,6 +395,55 @@ def test_ordering_grouping_by_key_transform(self):
             operator.itemgetter('key', 'count'),
         )
 
+    @skipIf(connection.vendor == 'oracle', "Oracle doesn't grouping by LOBs, see #24096.")
```
Suggestion (nit):

```diff
- Oracle doesn't grouping by LOBs, see #24096
+ Oracle cannot group by LOBs, see #24096
```

felixxm commented:
buildbot, test on oracle.

felixxm commented:
buildbot, test on oracle.

felixxm commented:
I fixed `Cast()` to `JSONField()` and removed unnecessary `DBMS_LOB.SUBSTR()` wrapping on Oracle.

felixxm commented:
I fixed `Cast()` to `JSONField()` on MariaDB.

pope1ni commented on tests/model_fields/test_jsonfield.py:
```diff
@@ -506,11 +506,11 @@ def test_none_key(self):
         if connection.vendor == 'oracle':
-            # On Oracle, the query returns JSON objects and arrays that do not have a 'null' value
-            # at the specified path, including those that do not have the key.
+            # Oracle support filtering JSON objects with NULL keys, but the
```
`support` → `supports`

laymonage replied:
I'd like to add, shouldn't it be `'null'` instead of `NULL`?

felixxm replied:
@laymonage No, it's a key value.

felixxm commented:
buildbot, test on oracle.

felixxm commented:
buildbot, test on selenium.

laymonage commented on docs/ref/models/fields.txt:
```diff
@@ -1175,6 +1175,74 @@ are converted to lowercase.
+    objects and arrays (represented in Python using :py:class:`dict` and
+    :py:class:`list`) are supported.
+
+.. _JSON1 extension: https://www.sqlite.org/json1.html
```
Do you think linking to `:ref:sqlite-json1` would be more useful? Right now it's not referenced in any other part of the docs.

felixxm replied:
Agreed.

felixxm commented:
Thanks y'all :rocket:

itdependsnetworks commented:
Been following just the last week or so, amazing job!!

laymonage commented:
Folks who want to use this on Django < 3.1 can look at [django-jsonfield-backport](https://github.com/laymonage/django-jsonfield-backport).

john-parton commented:
Referring to: https://code.djangoproject.com/ticket/33373

Is it possible to enable F() objects or expression on the RHS for HasKeys/HasAnyKeys just for the Postgres backend?

adamchainz commented:
Please don't comment on old pull requests.
